crypto/store/store_register.c

```diff
@@ -0,0 +1,226 @@
+{
+    loader->scheme = scheme;
+    return 1;
+}
+
+const char *STORE_LOADER_get0_scheme(const STORE_LOADER *loader)
+{
+    return loader->scheme;
+}
+
+int STORE_LOADER_set_open(STORE_LOADER *loader,
+                          STORE_open_fn store_open_function)
+{
+    loader->open = store_open_function;
+    return 1;
+}
+
+int STORE_LOADER_set_ctrl(STORE_LOADER *loader,
+                          STORE_ctrl_fn store_ctrl_function)
+{
+    loader->ctrl = store_ctrl_function;
+    return 1;
+}
+
+int STORE_LOADER_set_load(STORE_LOADER *loader,
+                          STORE_load_fn store_load_function)
+{
+    loader->load = store_load_function;
+    return 1;
+}
+
+int STORE_LOADER_set_eof(STORE_LOADER *loader,
+                         STORE_eof_fn store_eof_function)
+{
+    loader->eof = store_eof_function;
+    return 1;
+}
+
+int STORE_LOADER_set_close(STORE_LOADER *loader,
+                          STORE_close_fn store_close_function)
+{
+    loader->close = store_close_function;
+    return 1;
+}
+
+void STORE_LOADER_free(STORE_LOADER *loader)
+{
+    OPENSSL_free(loader);
+}
+
+/******************************************************************************
+ *
+ *  Functions for registering STORE_LOADERs
+ *
+ *****/
+
+static unsigned long store_loader_hash(const STORE_LOADER *v)
+{
+    return OPENSSL_LH_strhash(v->scheme);
+}
+
+static int store_loader_cmp(const STORE_LOADER *a, const STORE_LOADER *b)
+{
+    if (a->scheme != NULL && b->scheme != NULL)
+        return strcmp(a->scheme, b->scheme);
+    else if (a->scheme == b->scheme)
+        return 0;
+    return a->scheme == NULL ? -1 : 1;
+}
+
+static LHASH_OF(STORE_LOADER) *loader_register = NULL;
+
+int store_register_loader_int(STORE_LOADER *loader)
+{
+    const char *scheme = loader->scheme;
+
+    /*
+     * Check that the given scheme conforms to correct scheme syntax as per
+     * RFC 3986:
+     *
+     * scheme        = ALPHA *( ALPHA / DIGIT / "+" / "-" / "." )
+     */
+    if (isalpha(*scheme))
+        while (*scheme != '\0'
+               && (isalpha(*scheme)
+                   || isdigit(*scheme)
+                   || strchr("+-.", *scheme) != NULL))
+            scheme++;
+    if (*scheme != '\0') {
+        STOREerr(STORE_F_STORE_REGISTER_LOADER_INT, STORE_R_INVALID_SCHEME);
+        ERR_add_error_data(4, "scheme=", loader->scheme);
+        return 0;
+    }
+
+    if (loader_register == NULL) {
+        loader_register = lh_STORE_LOADER_new(store_loader_hash,
+                                              store_loader_cmp);
+        if (loader_register == NULL)
+            return 0;
+    }
+
+    if (lh_STORE_LOADER_insert(loader_register, loader) == NULL
+        && lh_STORE_LOADER_error(loader_register) > 0)
+        return 0;
+
+    return 1;
+}
+int STORE_register_loader(STORE_LOADER *loader)
+{
+    if (!store_init_once())
+        return 0;
+    return store_register_loader_int(loader);
```

Comment:
No really intentional, but post-poned for later consideration...  thanks for the reminder.